subsys/net/lib/tftp/tftp_client.c

```diff
@@ -0,0 +1,329 @@
+static u32_t  tftpc_buffer_size;
+
+/* TFTP Request block number property. */
+static u32_t  tftpc_block_no;
+static u32_t  tftpc_index;
+
+/* Global mutex to protect critical resources. */
+K_MUTEX_DEFINE(tftpc_lock);
+
+/* Global timeval structure indicating the timeout interval
+ * for all TFTP transactions.
+ */
+static struct timeval tftpc_timeout = {
+	.tv_sec   =   CONFIG_TFTPC_REQUEST_TIMEOUT,
+	.tv_usec  =   0
+};
+
+/* Number of simultaneous client connections will be
+ * NUM_FDS be minus 2
+ */
+static fd_set readfds;
+static int    fdnum;
+
+static int pollfds_add(int fd)
+{
+	FD_SET(fd, &readfds);
+
+	if (fd >= fdnum) {
+		fdnum = fd + 1;
+	}
+
+	return 0;
+}
+
+/* Name: make_request
+ * Description: This function takes in a given list of parameters and
+ * returns a read request packet. This packet can be sent
+ * out directly to the TFTP server.
+ */
+static u32_t make_request(const char *remote_file, const char *mode,
+						  u8_t request_type)
+{
+	u32_t      req_size;
+	const char def_mode[] = "octet";
+
+	/* Populate the read request with the provided params. Note that this
+	 * is created per RFC1350.
+	 */
+
+	/* Fill in the Request Type. Also keep tabs on the request size. */
+	put_u16(tftpc_buffer, request_type);
+	req_size = 2;
+
+	/* Copy in the name of the remote file - the file we need to get
+	 * from the server. Add an upper bound to ensure no buffers overflow.
+	 */
+	strncpy(tftpc_buffer + req_size, remote_file, TFTP_MAX_FILENAME_SIZE);
+	req_size += strlen(remote_file);
+
+	/* Fill in 0. */
+	tftpc_buffer[req_size] = 0x0;
+	req_size++;
+
+	/* Default to "Octet" if mode not specified. */
+	if (mode == NULL) {
+		mode = def_mode;
+	}
+
+	/* Copy the mode of operation. */
+	strncpy(tftpc_buffer + req_size, mode, TFTP_MAX_MODE_SIZE);
+	req_size += strlen(mode);
+
+	/* Fill in 0. */
+	tftpc_buffer[req_size] = 0x0;
+	req_size++;
+
+	return req_size;
+}
+
+/* Name: send_err
+ * Description: This function sends an Error report to the TFTP Server.
+ */
+static inline int send_err(int sock, int err_code, char *err_string)
+{
+	u32_t req_size;
+
+	LOG_ERR("Client Error. Sending code: %d(%s)", err_code, err_string);
+
+	/* Fill in the "Err" Opcode and the actual error code. */
+	put_u16(tftpc_buffer, ERROR_OPCODE);
+	put_u16(tftpc_buffer + 2, err_code);
+	req_size = 4;
+
+	/* Copy the Error String. */
+	strcpy(tftpc_buffer + req_size, err_string);
+	req_size += strlen(err_string);
+
+	/* Send Error to server. */
+	return send(sock, tftpc_buffer, req_size, 0);
+}
+
+/* Name: tftpc_recv_data
+ * Description: This function tries to get data from the TFTP Server
+ * (either response or data). Times out eventually.
+ */
+static int tftpc_recv_data(int sock)
+{
+	int     stat;
+
+	/* Enable select on the socket. */
+	pollfds_add(sock);
+
+	/* Enable select on this socket. Wait for the specified number
+	 * of seconds before exiting the call.
+	 */
+	stat = select(fdnum, &readfds, NULL, NULL, &tftpc_timeout);
```

Comment:
I would strongly recommend against use of select(). It's a legacy function, which was added to Zephyr solely to make it easier to port 3rd-party apps to run with Zephyr. But IMHO, nothing in Zephyr tree should use it in their implementation (besides tests/samples). Instead, poll() should be used.

cc @jukkar, I don't think we aware discussed this explicitly, so don't voting it down right away, but IMHO, we should have such a rule.

Reply:
> I don't think we aware discussed this explicitly, so don't voting it down right away, but IMHO, we should have such a rule.

Hmm, do not recall we have discussed this. But yes, the select() calls poll() so why not call poll() directly.

Reply:
> But yes, the select() calls poll() so why not call poll() directly.

Yes, in Zephyr, select() is implemented in terms of poll(), so using select() just adds more code and overhead for marshalling of fd_set <=> struct pollfd.

Reply:
Thanks for the suggestion.. This has removed a lot of unnecessary / redundant code from the implementation.. Incorporated and tested..